File: backend/app/schemas/project.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class ProjectCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=255)
    description: str | None = None


class ProjectUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None


class ProjectListParams(BaseModel):
    model_config = ConfigDict(extra="forbid")

    skip: int = Field(default=0, ge=0)
    limit: int = Field(default=20, ge=1, le=100)
    search: str | None = Field(default=None, min_length=1, max_length=255)
    sort: Literal[
        "created_at",
        "updated_at",
        "name",
        "-created_at",
        "-updated_at",
        "-name",
    ] = "-created_at"


class ProjectRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    description: str | None
    owner_id: UUID
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/project.ts
import type { PaginationParams } from "@/types/pagination";

export type Project = {
  id: string;
  name: string;
  description: string | null;
  owner_id: string;
  workspace_id: string;
  created_at: string;
  updated_at: string;
};

export type ProjectInput = {
  name: string;
  description: string | null;
};

export type ProjectSort =
  | "created_at"
  | "updated_at"
  | "name"
  | "-created_at"
  | "-updated_at"
  | "-name";

export type ProjectListParams = PaginationParams & {
  search?: string;
  workspace_id?: string;
  sort: ProjectSort;
};
